### PR TITLE
[WIP] Language Service hookup rewrite

### PR DESCRIPTION
**Pull request by davkean**
This is a rewrite of the language service hookup that moves from an all knowing single object `LanguageServiceHost` that is aware of all configurations and related IWorkspaceProjectContext, to instead individual configured-project-based services `WorkspaceContextHostInstance` that get fired up and torn down when a configuration becomes "implicitly" active. This naturally supports active config switches from debug -> release, and multi-targeting projects.

This also fixes: https://github.com/dotnet/project-system/issues/2733.

Currently this lives alongside the existing language service hookup, but is intended to replace it. 

Lots of TODOs:

- [ ] Feature flag to turn it on but turn off the other one
- [ ] Multi-targeting display names
- [ ] All IVsHierarchy interaction with Roslyn is through the "real" hierarchy and we handle any of the context properties
- [ ] "Active Context"
- [ ] All external iteration; Container Language, Error List, Edit & Continue, Code Model
- [ ] Need to push to UI thread
- [ ] Need to avoid overlapping evalution/design-time builds
- [ ] Prevent unloading project while updating context - but also bail early

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            _activationTracking = activationTracking;
+        }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            _activationTracking.ImplicitlyActivated += OnImplicitlyActivated;
+            _activationTracking.ImplicitlyDeactivated += OnImplicitlyDeactivated;
+
+            if (_activationTracking.IsImplicitlyActive)
+                return LoadAsync();
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs]: Is `LoadAsync` meant to be idempotent?  Isn't there a race here if the event fires between subscription and here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AnalyzerItemHandler.cs [hunk 2]
```diff
@@ -13,26 +13,37 @@ namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices.Handlers
...
+        private IWorkspaceProjectContext _context;
         private readonly HashSet<string> _paths = new HashSet<string>(StringComparers.Paths);
 
+        [ImportingConstructor]
+        public AnalyzerItemHandler(UnconfiguredProject project)
+            : this(project, null)
+        {
+        }
+
         [ImportingConstructor]
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AnalyzerItemHandler.cs]: Should be removed?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/ProjectPropertiesItemHandler.cs [hunk 3]
```diff
@@ -21,6 +21,16 @@ public ProjectPropertiesItemHandler(IWorkspaceProjectContext context)
...
         }
 
+        public string EvaluationRule
+        {
+            get { return ConfigurationGeneral.SchemaName; }
+        }
+
+        public void Initialize(IWorkspaceProjectContext context)
+        {
+            _context = context;
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/ProjectPropertiesItemHandler.cs]: It's expected that we can overwrite the once given in the constructor now?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/ProjectPropertiesItemHandler.cs]: Ah, problem is supporting both language services hosts; one grabs them via MEF, the other manually creates them. I'll add an assert.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs]: Yes it is.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AnalyzerItemHandler.cs (hunk 2 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AnalyzerItemHandler.cs]: Yes it should.

**Comment by davkean** [comment]
I had to rebase this, and will need to rebase in the future, so I'd hold off reviewing at the moment.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs]: There is a race here; if we deactivate after we check this property - we'll Load but never unload until project tear down.

**Comment by davkean** [comment]
Please still hold off reviewing this.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs]: Okay, I've figured out a better way to do this - I'm going to drop the events.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractImplicitlyActiveComponent.cs]: https://github.com/dotnet/project-system/pull/3689

**Comment by davkean** [comment]
This has been closed in lieu of the features/LanguageService branch.

**Comment by timbecn** [comment]
Hey @davkean - what does the closure of this ticket mean for #2733?

**Comment by Pilchie** [comment]
It's still being worked on, just one commit at a time in the `features/languageservice` branch instead of a single PR.

**Pull request closed without merging**